sql/schemachanger: reassign owned by and concurrent declarative schema changer drops can conflict

fqazi opened the pull request:
Informs: #87572

A number of bugs exist inside DROP OWNED / REASSIGNED OWNED BY inside the declarative schema changer:
1.  If a concurrent DROP operation for schema/database with a REASSIGNED OWNED BY could lead to the descriptors being dropped incorrectly by legacy schema changer jobs. This patch adds additional logic to ensure that only descriptors marked for drops are cleaned up.
2. The DROP OWNED BY support inside the declarative schema changer requires no privileges to be defined, unfortunately, this logic did not have proper error handling
3. When cleaning up back-references from schemas to the database, we skipped certain operations if the database was dropped, which is true after any cascaded DROP DATABASE statement phase. Which can lead to other issues depending on the stages executed.

fqazi committed:
sql: legacy schema changer jobs can incorrectly drop db/schema descriptors

Previously, the legacy schema changer would create a job
to clean up database/schema descriptors, however it would only
care if the descriptor was marked as dropped. This was
inadequate because with the declarative schema changer
becoming in use, its possible for it to interfere with
the declarative schema changer, if a legacy schema changer
job is kicked off after the pre-commit phase of a declarative
drop. To address this, the legacy schema changer now checks
if the descriptor ID is in the job details.

Release note: None

fqazi committed:
sql/schemachange: declarative schema changer could trap on drop owned by

Previously, we had logic to disable the declarative schema changer
for drop owned by if the user had any privileges in the
system.privilege table. Unfortunately, the error handling
for this logic was broken. This patch, properly checks
for errors in this code.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/93101)
<!-- Reviewable:end -->

fqazi committed:
sql/schemachanger: remove schemas from dropped descriptors

Previously, when a database descriptor was labeled as dropped,
we didn't bother cleaning up the schema entry inside. This could
be problematic in certain resolution code paths where we will
resolve databases by name even if they are even dropped, which
can cause these paths to hit a missing descriptor error. The
patch addresses, this bug by modifying dropped descriptors
inside the decalrative schema changer during execution.

Release note: None

postamar commented on pkg/sql/schema_changer.go:
```diff
@@ -694,7 +695,11 @@ func (sc *SchemaChanger) exec(ctx context.Context) error {
+				((desc.DescriptorType() == catalog.Database &&
+					sc.droppedDatabaseID == desc.GetID()) ||
+					(desc.DescriptorType() == catalog.Schema &&
+						sc.droppedSchemaIDs.Contains(desc.GetID()))) {
```
nit: consider simplifying this code by separating both cases:
```
// Some descriptors should be deleted if they are in the DROP state.
if !desc.Dropped() {
    return nil
}
switch desc.(type) {
    catalog.DatabaseDescriptor:
        if sc.droppedDatabaseID != desc.GetID() { return nil }
    case catalog.SchemaDescriptor:
        if !sc.droppedSchemaIDs.Constains(desc.GetID()) { return nil }
    default:
        return nil
}
_, err := sc.execCfg.DB.Del(ctx, catalogkeys.MakeDescMetadataKey(sc.execCfg.Codec, desc.GetID()))
return err
```

postamar reviewed (approved):
Nice!

fqazi commented:
@postamar @TFTR!

fqazi commented:
bors r+
